### PR TITLE
Introduce new preprocessor memory pool and with retries

### PR DESCRIPTION
**Pull request by glevkovich**
* **Problem Overview** *
1)
Introducing new Multi size buffer Pool for the pre-processor. Can be used as a utility. This pool allowes to save memory by choosing the right expected buffer and retrying in a rare case of failure. Most results in our system areless than 64K (99% of cases), so this is the 1st choice.
For more details - see MultiSizeBufferPool.hpp header.

2)
In addition, support for RawMemoryPool is kept: it is possible to choose which pool to use, or choose not to use any pool with the new parameter preProcessorMemoryPoolMode.

3) Fix il formed performance handler API. isRegisteredComponent doesn't make sense to have - since this is a time-of-check time-of-use issue (we are a multithreaded system). so I removed this API. The functions to register/unregister components should be  best effort. Nothing bad happens if I try to register/unregister the same component multiple times. so what? Tons of standard library APIs are defined like this. Example: set::insert, or unique_ptr::reset and the list is very long. No need to crash on this.

* **Testing Done**  
Added unuttests in MultiSizeBufferPool_test.cpp
Did leak checks.
Did Maestro LRT runs chess_plus and Freetrade.
 
no auto-bump

### CODE REVIEW

File: diagnostics/include/performance_handler.h [hunk 1]
```diff
@@ -353,9 +353,14 @@ struct HistogramData {
 
 class PerformanceHandler {
  public:
+  /*
+  If a component is already registered: print a warning and do nothing. This is the API definition, there is nothing
+  ill-defined here. For example std::vector::clear(), std::unique_ptr::reset() or std::set::insert() operates in the
+  same way. Same idea for the unregistered part.
+  */
```
Comment (yuliasherman) [inline diagnostics/include/performance_handler.h:360]: Please add to the PR description what you fixed in PerformanceHandler. Ideally, it should be a different PR.

**Review by yuliasherman — APPROVED** [review]
LGTM

File: diagnostics/include/performance_handler.h (hunk 1 above)
Comment (glevkovich) [reply diagnostics/include/performance_handler.h:360]: @yuliasherman , I agree it would have been better to have on another PR but I found this issue as part of my PR, some test crashed I beleive.
I've added a description on the above merge commit description.
Thanks.

File: bftengine/include/bftengine/ReplicaConfig.hpp [hunk 2]
```diff
@@ -261,10 +261,10 @@ class ReplicaConfig : public concord::serialize::SerializableFactory<ReplicaConf
...
   CONFIG_PARAM(enableEventGroups, bool, false, "A flag to specify whether event groups are enabled or not.");
 
-  CONFIG_PARAM(enablePreProcessorMemoryPool,
-               bool,
-               true,
-               "A flag to specify whether to use a memory pool in PreProcessor or not");
+  CONFIG_PARAM(preProcessorMemoryPoolMode,
+               uint32_t,
+               1,
+               "0: do not user a memory pool in PreProcessor, 1: use multi size buffer pool, 2: use raw memory pool");
```
Comment (Efrat1) [inline bftengine/include/bftengine/ReplicaConfig.hpp:267]: Did you mean "0: do not use"?

**Review by Efrat1 — APPROVED** [review]

File: bftengine/src/preprocessor/PreProcessor.hpp [hunk 3]
```diff
@@ -335,9 +339,18 @@ class PreProcessor : public bftEngine::IExternalObject {
   // Maximal subpool size should support up to 32MB (including responseSizeInternalOverhead_). So the maximal execution
   // engine supported size is kMaxSubpoolBuffersize_ - responseSizeInternalOverhead_
   static constexpr uint32_t kMaxSubpoolBuffersize_{1UL << 25};
-  // Must be sorted in an ascending order, by bufferSize
+
+  // Execution engine results can be stored in 3 different modes:
+  // 1) Without any pool, so the result buffers are managed by the pre-processor itself.
+  // 2) With MultiSizeBufferPool (default case)- the least amount of memory is consumed.
+  // 3) With MultiSizeBufferPool - legacy method. We keep this method to ensure stability - no abstract layer/interface
```
Comment (yuliasherman) [inline bftengine/src/preprocessor/PreProcessor.hpp:346]: Typo: -> RawMemoryPool

**Review by yuliasherman — APPROVED** [review]
LGTM

File: bftengine/include/bftengine/ReplicaConfig.hpp (hunk 2 above)
Comment (glevkovich) [reply bftengine/include/bftengine/ReplicaConfig.hpp:267]: Yes, strange - I fixed it myself. probably forgot to commit.
Thanks

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 4]
```diff
@@ -397,8 +394,21 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
       clientBatchingEnabled_(myReplica.getReplicaConfig().clientBatchingEnabled),
       threadPool_("PreProcessor::threadPool"),
       timers_{timers},
+      memoryPoolMode_{(myReplica_.getReplicaConfig().preProcessorMemoryPoolMode == 0)
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:397]: I'm going to move this into a function. I prefer not to initialize members in hpp on complex classes, it's not a godo habit in my opinion. members should be initialized in ctor.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 5]
```diff
@@ -397,8 +394,21 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
+                          ? PreProcessorMemoryPoolMode::NO_POOL
+                          : (myReplica_.getReplicaConfig().preProcessorMemoryPoolMode == 1)
+                                ? PreProcessorMemoryPoolMode::MULTI_SIZE_BUFFER_POOL
+                                : (myReplica_.getReplicaConfig().preProcessorMemoryPoolMode == 2)
+                                      ? PreProcessorMemoryPoolMode::RAW_MEMORY_POOL
+                                      : PreProcessorMemoryPoolMode::
+                                            NO_POOL /* should never happen and checked again later on ctor */},
       subpoolsConfig_{calcSubpoolsConfig()},
-      bufferPool_{createbufferPool()},
+      multiSizeBufferPool_{(memoryPoolMode_ == PreProcessorMemoryPoolMode::MULTI_SIZE_BUFFER_POOL)
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:406]: Why? multiSizeBufferPool_ is a unique_ptr. it is nullptr by default. always.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 6]
```diff
@@ -397,8 +394,21 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
+                               ? createMultiSizeBufferPool()
+                               : nullptr},
+      rawMemoryPool_{(memoryPoolMode_ == PreProcessorMemoryPoolMode::RAW_MEMORY_POOL)
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:409]: same as above, no need.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 7]
```diff
@@ -421,20 +431,31 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
       launchAsyncPreProcessJobRecorder_{histograms_.launchAsyncPreProcessJob},
       pm_{pm} {
-  // Validate subpoolsConfig_
-  if (!validateSubpoolsConfig()) {
+  clientMaxBatchSize_ = clientBatchingEnabled_ ? myReplica.getReplicaConfig().clientBatchingMaxMsgsNbr : 1;
+  const uint16_t numOfExternalClients = myReplica.getReplicaConfig().numOfExternalClients;
+  const uint16_t numOfReqEntries = numOfExternalClients * clientMaxBatchSize_;
+  const auto preProcessorMemoryPoolMode{myReplica_.getReplicaConfig().preProcessorMemoryPoolMode};
+
+  if (preProcessorMemoryPoolMode > 2) {
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:439]: I moved it to the initialize function already. So its in ctor.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 8]
```diff
@@ -472,7 +493,7 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
   RequestProcessingState::init(numOfRequiredReplies(), &histograms_);
   PreProcessReplyMsg::setPreProcessorHistograms(&histograms_);
   addTimers();
-}
+}  // namespace preprocessor
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:496]: That's a mistake, probably a auto-conflict mistake done by git.
Thanks for pointing out, removed it.

File: bftengine/src/preprocessor/PreProcessor.cpp [hunk 9]
```diff
@@ -421,20 +431,31 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
...
-  // Validate subpoolsConfig_
-  if (!validateSubpoolsConfig()) {
+  clientMaxBatchSize_ = clientBatchingEnabled_ ? myReplica.getReplicaConfig().clientBatchingMaxMsgsNbr : 1;
+  const uint16_t numOfExternalClients = myReplica.getReplicaConfig().numOfExternalClients;
+  const uint16_t numOfReqEntries = numOfExternalClients * clientMaxBatchSize_;
+  const auto preProcessorMemoryPoolMode{myReplica_.getReplicaConfig().preProcessorMemoryPoolMode};
+
+  if (preProcessorMemoryPoolMode > 2) {
+    std::ostringstream oss;
+    oss << "Invalid" << KVLOG(preProcessorMemoryPoolMode) << " ,must be in the range [0,2]!";
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.cpp:441]: Yes, we support 3 options 0,1 or 2.

File: bftengine/src/preprocessor/PreProcessor.hpp (hunk 3 above)
Comment (glevkovich) [reply bftengine/src/preprocessor/PreProcessor.hpp:346]: Thanks, fixed.

File: bftengine/src/preprocessor/PreProcessor.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuliasherman) [reply bftengine/src/preprocessor/PreProcessor.cpp:406]: Wrong place. I meant to initialize raw pointers to nullptr in .hpp instead of ctor.

**Review by eyalrund — APPROVED** [review]
